File: library/src/main/java/com/jdroid/java/domain/Entity.java
package com.jdroid.java.domain;

import java.io.Serializable;

/**
 * Domain Entity
 */
public abstract class Entity implements Serializable, Identifiable {

	private static final long serialVersionUID = 907671509045298947L;

	private String id;
	private String parentId;

	public Entity(String id) {
		this.id = id;
	}

	public Entity() {
	}

	/**
	 * Gets the identification for this {@link Entity}
	 *
	 * @return the id of this {@link Entity}
	 */
	@Override
	public String getId() {
		return id;
	}

	/**
	 * @param id the id to set
	 */
	public void setId(String id) {
		this.id = id;
	}

	public String getParentId() {
		return parentId;
	}

	/**
	 * @param parentId the parentId to set
	 */
	public void setParentId(String parentId) {
		this.parentId = parentId;
	}

	@Override
	public String toString() {
		StringBuilder builder = new StringBuilder();
		builder.append("Entity{id='");
		builder.append(id);
		builder.append("\', parentId='");
		builder.append(parentId);
		builder.append("\'}");
		return builder.toString();
	}

	/**
	 * Since equality has been redefined, so must be the hashCode function.
	 *
	 * @see java.lang.Object#hashCode()
	 */
	@Override
	public int hashCode() {
		final int prime = 31;
		return (id != null) ? prime * id.hashCode() : super.hashCode();
	}

	/**
	 * Redefines equality depending on the id of the entities being compared.
	 *
	 * @see java.lang.Object#equals(Object)
	 */
	@Override
	public boolean equals(Object obj) {
		if (obj == null) {
			return false;
		}
		if (!getClass().isAssignableFrom(obj.getClass())) {
			return false;
		}
		Entity other = (Entity)obj;

		if (id != null) {
			return id.equals(other.getId());
		}

		if (other.getId() != null) {
			return false;
		}

		return super.equals(other);
	}
}

File: repository/src/main/java/com/dipien/repository/Entity.kt
package com.dipien.repository

import java.io.Serializable

/**
 * Domain Entity
 */
abstract class Entity(private var id: String? = null) : Serializable, Identifiable {

    companion object {
        private const val serialVersionUID = 907671509045298947L
    }

    /**
     * @param parentId the parentId to set
     */
    var parentId: String? = null

    /**
     * Gets the identification for this [Entity]
     *
     * @return the id of this [Entity]
     */
    override fun getId(): String? {
        return id
    }

    /**
     * @param id the id to set
     */
    fun setId(id: String) {
        this.id = id
    }

    override fun toString(): String {
        val builder = StringBuilder()
        builder.append("Entity{id='")
        builder.append(id)
        if (parentId != null) {
            builder.append("\', parentId='")
            builder.append(parentId)
        }
        builder.append("\'}")
        return builder.toString()
    }

    /**
     * Since equality has been redefined, so must be the hashCode function.
     *
     * @see java.lang.Object.hashCode
     */
    override fun hashCode(): Int {
        val prime = 31
        return if (id != null) prime * id!!.hashCode() else super.hashCode()
    }

    /**
     * Redefines equality depending on the id of the entities being compared.
     *
     * @see java.lang.Object.equals
     */
    override fun equals(other: Any?): Boolean {
        if (other == null) {
            return false
        }
        if (!javaClass.isAssignableFrom(other.javaClass)) {
            return false
        }
        val otherEntity = other as Entity?

        if (id != null) {
            return id == otherEntity!!.getId()
        }

        return if (otherEntity!!.getId() != null) {
            false
        } else super.equals(otherEntity)
    }
}
